pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -591,10 +577,27 @@ func (s *ServiceEntryStore) edsUpdate(instances []*model.ServiceInstance, push b
 	for _, i := range instances {
 		keys[makeInstanceKey(i)] = struct{}{}
 	}
+	s.edsUpdateInSerial(keys, push)
+}
+
+// edsUpdateInSerial run s.edsUpdateByKeys in serial and wait for complete.
+func (s *ServiceEntryStore) edsUpdateInSerial(keys map[instancesKey]struct{}, push bool) {
+	// wait for the cache update finished
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	// trigger update eds endpoint shards
 	s.edsQueue.Push(func() error {
+		defer wg.Done()
 		s.edsUpdateByKeys(keys, push)
 		return nil
 	})
+	select {
+	// To prevent goroutine leak in tests.
+	case <-s.edsQueue.Closed():
+		return
+	default:
+		wg.Wait()
```

Comment:
I did see this leak in tests when running locally. Not sure why - maybe a flake?

Reply:
Maybe, the there is a race between wait and queue stopped. Need to fix

Reply:
fixed